userdata/userdata_cli_interop_test.go

```diff
@@ -0,0 +1,229 @@
+	Add      *NftCommand     `json:"add,omitempty"`
+	Flush    *NftCommand     `json:"flush,omitempty"`
+}
+
+type NftCli struct {
+	Nftables []NftCliObject `json:"nftables"`
+}
+
+func TestCommentInteropGo2Cli(t *testing.T) {
+	wantComment := "my comment"
+
+	c, err := nftables.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c.FlushRuleset()
+
+	table := c.AddTable(&nftables.Table{
+		Name:   "userdata-table",
+		Family: nftables.TableFamilyIPv4,
+	})
+
+	chain := c.AddChain(&nftables.Chain{
+		Name:  "userdata-chain",
+		Table: table,
+	})
+
+	c.AddRule(&nftables.Rule{
+		Table:    table,
+		Chain:    chain,
+		UserData: userdata.PutString(nil, userdata.TypeComment, wantComment),
+	})
+
+	if err := c.Flush(); err != nil {
+		t.Fatal(err)
+	}
+
+	out := bytes.NewBuffer(nil)
+	d := exec.Command("nft", "-j", "list", "table", "userdata-table")
+	d.Stdout = out
+	if err := d.Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	var outJson NftCli
+	if err := json.Unmarshal(out.Bytes(), &outJson); err != nil {
+		t.Fatal()
+	}
+
+	found := 0
+	for _, e := range outJson.Nftables {
+		if e.Rule == nil || e.Rule.Handle == 0 {
+			continue
+		}
+
+		if e.Rule.Comment != wantComment {
+			t.Fatal()
+		}
+
+		found++
+	}
+
+	if found != 1 {
+		t.Fatalf("found %d rules", found)
+	}
+
+	c.DelTable(table)
+
+	if err := c.Flush(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestCommentInteropCli2Go(t *testing.T) {
+	wantComment := "my comment"
+
+	inJson := NftCli{
+		Nftables: []NftCliObject{
+			{
+				Metainfo: &NftCliMetainfo{
+					JSONSchemaVersion: 1,
+				},
+			},
+			{
+				Flush: &NftCommand{
+					Ruleset: nil,
+				},
+			},
+			{
+				Add: &NftCommand{
+					Table: &NftCliTable{
+						Family: "ip",
+						Name:   "userdata-table",
+					},
+				},
+			},
+			{
+				Add: &NftCommand{
+					Chain: &NftCliChain{
+						Family: "ip",
+						Name:   "userdata-chain",
+						Table:  "userdata-table",
+					},
+				},
+			},
+			{
+				Add: &NftCommand{
+					Rule: &NftCliRule{
+						Family:  "ip",
+						Table:   "userdata-table",
+						Chain:   "userdata-chain",
+						Comment: wantComment,
+						Expr:    []NftCliExpr{},
+					},
+				},
+			},
+		},
+	}
+
+	in := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(in).Encode(inJson); err != nil {
+		t.Fatal()
+	}
+
+	d := exec.Command("nft", "-j", "-f", "-")
+	d.Stdin = in
+	if err := d.Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	c, err := nftables.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	table := &nftables.Table{
+		Name:   "userdata-table",
+		Family: nftables.TableFamilyIPv4,
+	}
+
+	chain := &nftables.Chain{
+		Name:  "userdata-chain",
+		Table: table,
+	}
+
+	rules, err := c.GetRules(table, chain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(rules) != 1 {
+		t.Fatal()
+	}
+
+	if comment, ok := userdata.GetString(rules[0].UserData, userdata.TypeComment); !ok {
+		t.Fatalf("failed to find comment")
+	} else if comment != wantComment {
+		t.Fatalf("comment mismatch %s != %s", comment, wantComment)
```

Comment:
Use %q instead of %s here so that special characters are easily recognizeable

Reply:
Done